src/chips/hotel/src/gpio.rs

```diff
@@ -56,29 +85,64 @@ pub enum Pin {
     P15,
 }
 
-#[derive(Debug)]
-pub struct GPIOPin {
+pub struct Pin {
     port: *mut PortRegisters,
-    pin: Pin,
+    pin: PinNum,
+    client_data: Cell<usize>, 
+    change: Cell<bool>, 
+    client: TakeCell<&'static Client>,
 }
 
-impl GPIOPin {
-    const fn new(port: *mut PortRegisters, pin: Pin) -> GPIOPin {
-        GPIOPin {
+impl Pin {
+    const fn new(port: *mut PortRegisters, pin: PinNum) -> Pin {
+        Pin {
             port: port,
             pin: pin,
+            change: Cell::new(false),
+            client_data: Cell::new(0),
+            client: TakeCell::empty(),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let mask = 1 << (self.pin as u32);
+
+        let port: &mut PortRegisters = unsafe { transmute(self.port) };
+        port.interrupt_status.set(mask);
+
+        // If our InterruptMode was `Change`, we need to flip the direction of
+        // the interrupt polarity.
+        if self.change.get() {
+            if port.interrupt_pol_set.get() & mask != 0 {
+                port.interrupt_pol_clear.set(mask);
+            } else {
+                port.interrupt_pol_set.set(mask);
+            }
         }
+
+        self.client.map(|client| {
+            client.fired(self.client_data.get());
+        });
+    }
+
+    pub fn set_client(&self, client: &'static Client) {
+        self.client.put(Some(client));
     }
 }
 
-impl hil::gpio::GPIOPin for GPIOPin {
+impl GPIOPin for Pin {
     fn enable_output(&self) {
         let port: &mut PortRegisters = unsafe { transmute(self.port) };
         port.output_enable.set(1 << (self.pin as u32));
     }
 
-    fn enable_input(&self, _mode: hil::gpio::InputMode) {
-        // TODO(alevy): implement
+    fn enable_input(&self, _mode: InputMode) {
+        // Noop, input is always enabled on this chip
+
+        // InputMode equivilant is actually set in the Pinmux. It actually kind of
+        // makes sense to have this be something that's setup by the platform
+        // initialization, rather than chosen by the client of a particular GPIO
+        // pin.
```

Comment:
This is a bummer. I've discussed with @ppannuto and maybe this implies that the GPIO hil just shouldn't imply you can change the pull state of the pin. Instead a different trait would do that. On the SAM4L it would happen to be implemented by the same controller, on others, like this one, not. The system call interface wouldn't have to change. For now, this seems like a rare enough case that i don't think it makes sense to pull pinmux into this module (even though we could, technically).